<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.18.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>komet module &mdash; komet 1.0 documentation</title>
      <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
    <link rel="shortcut icon" href="_static/favicon.png"/>
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script src="_static/jquery.js"></script>
        <script src="_static/_sphinx_javascript_frameworks_compat.js"></script>
        <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
        <script src="_static/doctools.js"></script>
        <script src="_static/sphinx_highlight.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="prev" title="Komet with Uniprot LAkernel" href="vignettes/komet_uniprotLAkernel.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="index.html" class="icon icon-home">
            komet
              <img src="_static/komet-logo-small.png" class="logo" alt="Logo"/>
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Getting started</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="vignettes/komet.html">komet : Kronecker Optimized METhod for DTI prediction</a></li>
<li class="toctree-l1"><a class="reference internal" href="vignettes/komet_uniprotLAkernel.html">Komet with Uniprot LAkernel</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">API</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="current reference internal" href="#">komet module</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#komet.Morgan_FP"><code class="docutils literal notranslate"><span class="pre">Morgan_FP()</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="#komet.Nystrom_X"><code class="docutils literal notranslate"><span class="pre">Nystrom_X()</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="#komet.Nystrom_X_cn"><code class="docutils literal notranslate"><span class="pre">Nystrom_X_cn()</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="#komet.SVM_bfgs"><code class="docutils literal notranslate"><span class="pre">SVM_bfgs()</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="#komet.add_indfasta"><code class="docutils literal notranslate"><span class="pre">add_indfasta()</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="#komet.add_indsmiles"><code class="docutils literal notranslate"><span class="pre">add_indsmiles()</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="#komet.compute_proba"><code class="docutils literal notranslate"><span class="pre">compute_proba()</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="#komet.compute_proba_Platt_Scalling"><code class="docutils literal notranslate"><span class="pre">compute_proba_Platt_Scalling()</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="#komet.load_datas"><code class="docutils literal notranslate"><span class="pre">load_datas()</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="#komet.load_df"><code class="docutils literal notranslate"><span class="pre">load_df()</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="#komet.make_train_test_val_S1"><code class="docutils literal notranslate"><span class="pre">make_train_test_val_S1()</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="#komet.make_train_test_val_S2"><code class="docutils literal notranslate"><span class="pre">make_train_test_val_S2()</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="#komet.make_train_test_val_S3"><code class="docutils literal notranslate"><span class="pre">make_train_test_val_S3()</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="#komet.make_train_test_val_S4"><code class="docutils literal notranslate"><span class="pre">make_train_test_val_S4()</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="#komet.process_LCIdb"><code class="docutils literal notranslate"><span class="pre">process_LCIdb()</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="#komet.results"><code class="docutils literal notranslate"><span class="pre">results()</span></code></a></li>
</ul>
</li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">komet</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">komet module</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/komet.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="module-komet">
<span id="komet-module"></span><h1>komet module<a class="headerlink" href="#module-komet" title="Permalink to this heading"></a></h1>
<dl class="py function">
<dt class="sig sig-object py" id="komet.Morgan_FP">
<span class="sig-prename descclassname"><span class="pre">komet.</span></span><span class="sig-name descname"><span class="pre">Morgan_FP</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">list_smiles</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/komet.html#Morgan_FP"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#komet.Morgan_FP" title="Permalink to this definition"></a></dt>
<dd><p>Computes the Morgan fingerprints for a list of SMILES strings.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>list_smiles</strong> (<em>list</em>) – A list of SMILES strings for which to compute the fingerprints.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A tensor containing the Morgan fingerprints for the input SMILES strings.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>torch.Tensor</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="komet.Nystrom_X">
<span class="sig-prename descclassname"><span class="pre">komet.</span></span><span class="sig-name descname"><span class="pre">Nystrom_X</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">smiles_list</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">S</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">MorganFP</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">V</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rM</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">Mu</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">epsi</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/komet.html#Nystrom_X"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#komet.Nystrom_X" title="Permalink to this definition"></a></dt>
<dd><p>Computes the approximate features of the molecular kernel using the Nystrom approximation.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>smiles_list</strong> – A list of SMILES strings for which to compute features.</p></li>
<li><p><strong>S</strong> – Indices of the subset of samples used for the Nystrom approximation.</p></li>
<li><p><strong>MorganFP</strong> – Precomputed Morgan Fingerprints for the dataset.</p></li>
<li><p><strong>V</strong> – Eigenvectors of the kernel matrix.</p></li>
<li><p><strong>rM</strong> – Rank of the approximation.</p></li>
<li><p><strong>Mu</strong> – Eigenvalues of the kernel matrix.</p></li>
<li><p><strong>epsi</strong> – Small regularization term added for numerical stability.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Approximated feature matrix for the input SMILES list.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>torch.Tensor</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="komet.Nystrom_X_cn">
<span class="sig-prename descclassname"><span class="pre">komet.</span></span><span class="sig-name descname"><span class="pre">Nystrom_X_cn</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">mM</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rM</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nM</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">MorganFP</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/komet.html#Nystrom_X_cn"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#komet.Nystrom_X_cn" title="Permalink to this definition"></a></dt>
<dd><p>Compute the Nystrom approximation of the centered normalized feature matrix.</p>
<dl class="field-list">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>mM</strong> (<em>int</em>) – Number of molecule landmarks</p></li>
<li><p><strong>rM</strong> (<em>int</em>) – Number of dimensions to keep after SVD for molecule features</p></li>
<li><p><strong>nM</strong> (<em>int</em>) – Total number of molecules.</p></li>
<li><p><strong>MorganFP</strong> (<em>numpy.ndarray</em>) – Matrix of Morgan fingerprints of shape (nM, fingerprint_length).</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The centered normalized feature matrix.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>torch.Tensor</p>
</dd>
<dt class="field-even">Notes<span class="colon">:</span></dt>
<dd class="field-even"><p>This function computes the Nystrom approximation of the feature matrix using the given
Morgan fingerprints. It first selects a random subset S of size mM from the total
nM molecules. It then computes the kernel matrix K using the selected subset.
The approximate feature matrix is computed using Singular Value Decomposition (SVD)
on K. Finally, the features are normalized by centering and dividing by their L2 norm.</p>
<p>The input MorganFP should be a numpy array with shape (nM, fingerprint_length).</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="komet.SVM_bfgs">
<span class="sig-prename descclassname"><span class="pre">komet.</span></span><span class="sig-name descname"><span class="pre">SVM_bfgs</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">X_cn</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">Y_cn</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">y</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">I</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">J</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">lamb</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">niter</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">50</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/komet.html#SVM_bfgs"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#komet.SVM_bfgs" title="Permalink to this definition"></a></dt>
<dd><p>Trains an SVM model using the L-BFGS optimization algorithm to minimize the loss function.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>X_cn</strong> – Feature matrix for compounds.</p></li>
<li><p><strong>Y_cn</strong> – Feature matrix for targets.</p></li>
<li><p><strong>y</strong> – Interaction labels.</p></li>
<li><p><strong>I</strong> – Indices of molecules.</p></li>
<li><p><strong>J</strong> – Indices of proteins.</p></li>
<li><p><strong>lamb</strong> – Regularization parameter.</p></li>
<li><p><strong>niter</strong> – Number of iterations for the L-BFGS optimization algorithm.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Optimized weight and bias parameters for the SVM model, and the history of the loss function.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>tuple[torch.Tensor, torch.Tensor, list[float]]</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="komet.add_indfasta">
<span class="sig-prename descclassname"><span class="pre">komet.</span></span><span class="sig-name descname"><span class="pre">add_indfasta</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/komet.html#add_indfasta"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#komet.add_indfasta" title="Permalink to this definition"></a></dt>
<dd><p>Adds an index column for each unique FASTA sequence in the dataframe.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>df</strong> – Dataframe containing protein sequences.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The updated dataframe with an ‘indfasta’ column and the array of unique FASTA sequences.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>tuple[pd.DataFrame, np.ndarray]</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="komet.add_indsmiles">
<span class="sig-prename descclassname"><span class="pre">komet.</span></span><span class="sig-name descname"><span class="pre">add_indsmiles</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/komet.html#add_indsmiles"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#komet.add_indsmiles" title="Permalink to this definition"></a></dt>
<dd><p>Adds a column to the dataframe with indices for each unique SMILES string.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>df</strong> (<em>pd.DataFrame</em>) – The dataframe to be processed.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A tuple containing the processed dataframe with an added column for SMILES indices
and the array of unique SMILES strings.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>tuple(pd.DataFrame, np.ndarray)</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="komet.compute_proba">
<span class="sig-prename descclassname"><span class="pre">komet.</span></span><span class="sig-name descname"><span class="pre">compute_proba</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">w_bfgs</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">b_bfgs</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">s</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">t</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">X_cn</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">Y_cn</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">I</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">J</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/komet.html#compute_proba"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#komet.compute_proba" title="Permalink to this definition"></a></dt>
<dd><p>Computes probabilities using the trained weights and Platt scaling parameters.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>w_bfgs</strong> – Optimized weights from the SVM model for compounds.</p></li>
<li><p><strong>b_bfgs</strong> – Optimized bias from the SVM model.</p></li>
<li><p><strong>s</strong> – Optimized scaling parameter from Platt scaling.</p></li>
<li><p><strong>t</strong> – Optimized offset parameter from Platt scaling.</p></li>
<li><p><strong>X_cn</strong> – Feature matrix for compounds.</p></li>
<li><p><strong>Y_cn</strong> – Feature matrix for targets.</p></li>
<li><p><strong>I</strong> – Indices of molecules.</p></li>
<li><p><strong>J</strong> – Indices of proteins.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A tuple containing the margin values, predicted labels, and predicted probabilities.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>tuple[torch.Tensor, torch.Tensor, torch.Tensor]</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="komet.compute_proba_Platt_Scalling">
<span class="sig-prename descclassname"><span class="pre">komet.</span></span><span class="sig-name descname"><span class="pre">compute_proba_Platt_Scalling</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">w_bfgs</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">X_cn</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">Y_cn</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">y</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">I</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">J</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">niter</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">20</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/komet.html#compute_proba_Platt_Scalling"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#komet.compute_proba_Platt_Scalling" title="Permalink to this definition"></a></dt>
<dd><p>Computes probability estimates for interaction predictions using Platt scaling.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>w_bfgs</strong> – Optimized weights from the SVM model.</p></li>
<li><p><strong>X_cn</strong> – Feature matrix for compounds.</p></li>
<li><p><strong>Y_cn</strong> – Feature matrix for targets.</p></li>
<li><p><strong>y</strong> – Interaction labels.</p></li>
<li><p><strong>I</strong> – Indices of molecules.</p></li>
<li><p><strong>J</strong> – Indices of proteins.</p></li>
<li><p><strong>niter</strong> – Number of iterations for the L-BFGS optimization algorithm.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Optimized parameters ‘s’ and ‘t’ for Platt scaling, and the history of the loss function.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>tuple[torch.Tensor, torch.Tensor,list[float]]</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="komet.load_datas">
<span class="sig-prename descclassname"><span class="pre">komet.</span></span><span class="sig-name descname"><span class="pre">load_datas</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/komet.html#load_datas"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#komet.load_datas" title="Permalink to this definition"></a></dt>
<dd><p>Loads interaction data including indices of molecule and protein pairs along with their labels.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>df</strong> – Dataframe containing the interaction data.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Tensors of molecule indices, protein indices, and interaction labels.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>tuple[torch.Tensor, torch.Tensor, torch.Tensor]</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="komet.load_df">
<span class="sig-prename descclassname"><span class="pre">komet.</span></span><span class="sig-name descname"><span class="pre">load_df</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">name</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dataset_dir</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'data'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/komet.html#load_df"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#komet.load_df" title="Permalink to this definition"></a></dt>
<dd><p>Loads a dataframe from a CSV file, cleans up SMILES strings that cannot be read by RDKit,
and returns the cleaned dataframe.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>name</strong> (<em>str</em>) – The name of the file (with extension) to be loaded. If the file is a zip archive,
it will be extracted first.</p></li>
<li><p><strong>dataset_dir</strong> (<em>str</em><em>, </em><em>optional</em>) – The directory where the dataset is located, by default “data”.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The cleaned dataframe with SMILES strings that RDKit can read.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pd.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="komet.make_train_test_val_S1">
<span class="sig-prename descclassname"><span class="pre">komet.</span></span><span class="sig-name descname"><span class="pre">make_train_test_val_S1</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">train_ratio</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.8</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">test_ratio</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.12</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/komet.html#make_train_test_val_S1"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#komet.make_train_test_val_S1" title="Permalink to this definition"></a></dt>
<dd><p>Splits the input DataFrame into training, testing, and validation datasets. The function first converts the DataFrame into a
numpy matrix where rows correspond to ‘indfasta’ (protein indices), columns to ‘indsmiles’ (drug indices), and cell values to ‘score’ (interaction score).
It then identifies positive (interaction score = 1) and negative (interaction score = 0) interactions and distributes them into
training, testing, and validation sets according to the specified ratios. The split is done ensuring that each set contains a
balanced proportion of positive and unknown interactions.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df</strong> (<em>pandas.DataFrame</em>) – Input data containing the columns ‘indfasta’, ‘indsmiles’, and ‘score’.</p></li>
<li><p><strong>train_ratio</strong> (<em>float</em><em>, </em><em>optional</em>) – The proportion of the dataset to be used for the training set, defaults to 0.8.</p></li>
<li><p><strong>test_ratio</strong> (<em>float</em><em>, </em><em>optional</em>) – The proportion of the dataset to be used for the testing set, defaults to 0.12.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A tuple of numpy arrays representing the training, testing, and validation datasets respectively.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>(numpy.ndarray, numpy.ndarray, numpy.ndarray)</p>
</dd>
</dl>
<p>Note:
The remaining portion of the dataset not allocated to training or testing is used for validation.
This function assumes that the DataFrame’s ‘score’ column contains binary values (1 for interaction and 0 for no interaction).
NaN values in ‘score’ are treated as unknown interactions and are handled separately.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="komet.make_train_test_val_S2">
<span class="sig-prename descclassname"><span class="pre">komet.</span></span><span class="sig-name descname"><span class="pre">make_train_test_val_S2</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/komet.html#make_train_test_val_S2"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#komet.make_train_test_val_S2" title="Permalink to this definition"></a></dt>
<dd><p>Splits the interaction data into distinct training, testing, and validation datasets,
ensuring molecules in the test set are not in the training set and molecules in the validation set
are not in the training set. It processes input data to form interaction matrices and categorizes
these interactions into positive, negative, and unknown based on their presence, absence, or uncertainty in the dataset.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>df</strong> (<em>pandas.DataFrame</em>) – The input data containing the columns ‘indfasta’, ‘indsmiles’, and ‘score’, representing interaction data between proteins and molecules.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A tuple of numpy.ndarrays representing the training, testing, and validation datasets. Each set contains arrays of interactions labeled with indices and interaction scores, with drugs and proteins split according to specified criteria to ensure separation between training, testing, and validation sets.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>tuple</p>
</dd>
</dl>
<p>The function performs the following steps:
- Convert the DataFrame to a numpy array of interaction scores.
- Determine the unique proteins and drugs, assigning them to train, test, and validation groups based on specified proportions.
- For each group, create datasets of positive and negative interactions, handling missing values as unknown interactions.
- Ensure there is no overlap of molecules between the training set and either the testing or validation sets.
- Return the prepared datasets for further processing or model training.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="komet.make_train_test_val_S3">
<span class="sig-prename descclassname"><span class="pre">komet.</span></span><span class="sig-name descname"><span class="pre">make_train_test_val_S3</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/komet.html#make_train_test_val_S3"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#komet.make_train_test_val_S3" title="Permalink to this definition"></a></dt>
<dd><p>Splits the interaction matrix into training, testing, and validation sets, ensuring there is no overlap
between the proteins in the training set and those in the testing or validation sets. The split is based
on the initial distribution of proteins and molecules in the interaction matrix.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>df</strong> (<em>pandas.DataFrame</em>) – The input DataFrame containing the columns ‘indfasta’ for proteins, ‘indsmiles’ for molecules, and ‘score’ for their interaction scores.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A tuple containing the training, testing, and validation sets, each as a numpy.ndarray with the first column representing the index of proteins, the second column the index of molecules, and the third column the interaction scores (1 for interaction, 0 for no interaction, and np.nan for unknown interactions).</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>tuple</p>
</dd>
</dl>
<p>The function performs the following operations:
- Converts the DataFrame to a numpy array representing the interaction scores.
- Randomly shuffles and splits proteins and molecules into distinct groups for training, testing, and validation based on predefined ratios.
- Creates interaction datasets for each set, ensuring proteins in the test set are not in the train set, and proteins in the val set are not in the train set, thereby preventing data leakage.
- Handles missing values and ensures the final datasets are balanced in terms of positive and negative interactions.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="komet.make_train_test_val_S4">
<span class="sig-prename descclassname"><span class="pre">komet.</span></span><span class="sig-name descname"><span class="pre">make_train_test_val_S4</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/komet.html#make_train_test_val_S4"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#komet.make_train_test_val_S4" title="Permalink to this definition"></a></dt>
<dd><p>Splits the interaction matrix into balanced training, testing, and validation sets ensuring there is no
overlap between the proteins and molecules in the training set compared to those in the testing and
validation sets. This aims to create distinct and balanced sets for robust model evaluation.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>df</strong> (<em>pandas.DataFrame</em>) – Input DataFrame containing interaction data with ‘indfasta’ for proteins, ‘indsmiles’ for molecules,
and ‘score’ for interaction scores.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A tuple containing the training, testing, and validation datasets. Each dataset consists of a numpy array with three columns: protein indices, molecule indices, and interaction scores. The interaction scores are 1 for positive interactions, 0 for negative interactions, and np.nan for unknown interactions.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>tuple</p>
</dd>
</dl>
<p>The function executes the following major steps:
- Converts the DataFrame to a numpy array to represent the interaction scores.
- Splits proteins and molecules into distinct groups for training, testing, and validation based on predefined criteria, ensuring there’s no overlap between the sets for proteins and molecules.
- Constructs interaction datasets for each set by maintaining a balance between positive and negative interactions and properly handling unknown interactions.
- Ensures the testing and validation sets are balanced and do not contain any proteins or molecules present in the training set, thus avoiding data leakage and ensuring the model’s generalizability.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="komet.process_LCIdb">
<span class="sig-prename descclassname"><span class="pre">komet.</span></span><span class="sig-name descname"><span class="pre">process_LCIdb</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">name_file</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">data_dir</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'./'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_length_fasta</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1000</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bioactivity_choice</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'checkand1database'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_weight</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">100</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_weight</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">900</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">interaction_plus</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1e-07</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">interaction_minus</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.0001</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/komet.html#process_LCIdb"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#komet.process_LCIdb" title="Permalink to this definition"></a></dt>
<dd><p>Processes data from a given ligand-chemical interaction database file and performs various data cleaning and transformation steps.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>name_file</strong> (<em>str</em>) – The name of the file containing the database information.</p></li>
<li><p><strong>data_dir</strong> (<em>str</em><em>, </em><em>optional</em>) – The directory where the database file and the file prot_uniprot_fasta_all.csv are located, defaults to “./”.</p></li>
<li><p><strong>max_length_fasta</strong> (<em>int</em><em>, </em><em>optional</em>) – The maximum length of FASTA sequences to be included, defaults to 1000.</p></li>
<li><p><strong>bioactivity_choice</strong> (<em>str</em><em>, </em><em>optional</em>) – The type of bioactivity data to include, defaults to ‘checkand1database’.</p></li>
<li><p><strong>min_weight</strong> (<em>int</em><em>, </em><em>optional</em>) – The minimum molecular weight of the ligands to be included, defaults to 100.</p></li>
<li><p><strong>max_weight</strong> (<em>int</em><em>, </em><em>optional</em>) – The maximum molecular weight of the ligands to be included, defaults to 900.</p></li>
<li><p><strong>interaction_plus</strong> (<em>float</em><em>, </em><em>optional</em>) – The threshold for positive interaction, defaults to 1e-7.</p></li>
<li><p><strong>interaction_minus</strong> (<em>float</em><em>, </em><em>optional</em>) – The threshold for negative interaction, defaults to 1e-4.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A pandas DataFrame containing the processed and filtered data.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pandas.DataFrame</p>
</dd>
</dl>
<p>The function performs the following steps:
- Reads the database and filters based on activity type (pIC50, pKd, pKi) and units (negative logarithm).
- Merges with protein FASTA sequence data and filters based on sequence length and molecular weight.
- Processes bioactivity data based on the selected type and calculates mean, min, and max bioactivity values.
- Aggregates data based on SMILES representation and FASTA sequence, and calculates interaction scores.
- Saves the processed data to a new CSV file.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="komet.results">
<span class="sig-prename descclassname"><span class="pre">komet.</span></span><span class="sig-name descname"><span class="pre">results</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">y</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">y_pred</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">proba_pred</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/komet.html#results"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#komet.results" title="Permalink to this definition"></a></dt>
<dd><p>Computes and returns various performance metrics of the model including accuracy, AUC, average precision,
optimal threshold for classification, accuracy at the best threshold, confusion matrix, and false positive rate.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>y</strong> – True labels.</p></li>
<li><p><strong>y_pred</strong> – Predicted labels before applying Platt scaling.</p></li>
<li><p><strong>proba_pred</strong> – Predicted probabilities after applying Platt scaling.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A tuple containing accuracy before Platt scaling, AUC, average precision, optimal threshold,
accuracy at optimal threshold, confusion matrix, and false positive rate at the optimal threshold.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>tuple[float, float, float, float, float, np.ndarray, float]</p>
</dd>
</dl>
</dd></dl>

</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="vignettes/komet_uniprotLAkernel.html" class="btn btn-neutral float-left" title="Komet with Uniprot LAkernel" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2024, Gwenn Guichaoua.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>